ENH: add `contract`: optimizing numpy's einsum expression

dgasmith opened the pull request:
Greetings everyone,
Numpy's einsum function can compute arbitrary expressions; however, there are two drawbacks to using pure einsum: einsum does not consider building intermediate arrays for possible reductions in overall rank and is not currently capable of using a vendor BLAS. The `contract` function aims to solve these issues and effectively set out to answer the question: "How fast can I execute a Einstein summation expression in numpy?" The answer turns out to be quite fast, often only 2-3 times slower than an optimized C or Fortran expression, especially when a vendor BLAS utilized. In addition, multi-argument einsum speed issues (like https://github.com/numpy/numpy/issues/5366) should be solved and, when BLAS functionality is added, can reproduce multi_dot like behavior (https://github.com/numpy/numpy/pull/4977).

The original [repository](https://github.com/dgasmith/opt_einsum) has additional information that I have not been able to cram into the current documentation. There are additional test scripts that provide random expression testing, single expression debugging, and path comparisons, in addition to evaluate timing. Other branches on this repo have both BLAS and ellipses in the subscript functionality.

Following a few recent PR's I have decided to break this up into three major commits:
- Initial commit. The basics; input parsing, both path functions, and the necessary equipment for looping through the contraction pairs.
- BLAS commit: Adds BLAS functionality. This can either be done by using tensordot (tensordot would need an update to prevent arbitrary copies) or by building tensordot like functionality.
- Index commit: Order einsum ouput indices to try and reduce the number of nd-transposes for BLAS and try to build indices that are optimal for einsum (poor einsum indexing order can result in 2x slowdowns).

I think this will help split up some of the complexity and the first PR is already large enough.

jaimefrio commented on numpy/linalg/linalg.py:
```diff
@@ -2343,3 +2343,473 @@ def _multi_dot(arrays, order, i, j):
+       5     False            bcde,fb->cdef                         gc,hd,cdef->efgh
+       5     False            cdef,gc->defg                            hd,defg->efgh
+       5     False            defg,hd->efgh                               efgh->efgh
+    >>> ein_result = np.einsum('ea,fb,abcd,gc,hd->efgh', C, C, I, C, C, path=opt_path[0])
```
There is no line calculating `opt_result`, and this one computing `ein_result` is passing a wrong keyword argument `pat` to `np.einsum`.

jaimefrio commented on numpy/linalg/linalg.py:
```diff
@@ -2343,3 +2343,473 @@ def _multi_dot(arrays, order, i, j):
+    """
+
+    # Parse input
+    if not isinstance(subscripts, str):
```
This is OK for starters, but `np.einsum` supports an alternate method of providing the input that should probably also be supported, see [the docs](http://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum.html).

dgasmith replied:
I have previously implemented ellipses in python and it was fairly unwieldy. If this should support einsum subscript syntax identical to that of np.einsum itself I wonder if this function should just leverage that [machinery](https://github.com/numpy/numpy/blob/e0807a56f6c6717b006ba80c86c3149d51990f63/numpy/core/src/multiarray/multiarraymodule.c#L2612).

shoyer commented:
For the opt_path option, it would be nice to return a machine readable version, i.e., something that I could paste in to substitute for `einsum`. It would be even better if this function used a cache for its arguments (sort of like numba's `jit`) so the optimal path for arguments with fixed shape only needs to be calculated once.

My vote is also for a name like `einsumopt` to make clear the shared origin and differences from `einsum`. `contract` sounds very generic to me.

dgasmith commented:
@shoyer 
For machine readable are you thinking something along the lines of:

```python
# Regular einsum
np.einsum('aab,bc', op1, op2)
# Machine readable path return
np.dot(np.einsum('aab->ab', op1), op2)
```

This can become exceedingly long for the more complicated paths. I have written cache like algorithms for my own usage of this type of script, I will look into adding them here.

I was thinking `contract` because it should be a mixing pot of different algorithms. Right now it only uses `einsum`; however, it will use vendor BLAS before it is complete and anything else that comes along can be added as well. For example, if numpy ever offered an interface to a tensor framework this could be leveraged as well.

njsmith commented:
Shouldn't the name be "einsum"? Names are for users, and users care about
interfaces, not internal implementation. I don't see why we should have two
different public functions that do the same thing?
On 3 Feb 2015 06:13, "Daniel Smith" <notifications@github.com> wrote:

> @shoyer <https://github.com/shoyer>
> For machine readable are you thinking something along the lines of:
>
> # Regular einsum
> np.einsum('aab,bc', op1, op2)# Machine readable path return
> np.dot(np.einsum('aab->ab', op1), op2)
>
> This can become exceedingly long for the more complicated paths. I have
> written cache like algorithms for my own usage of this type of script, I
> will look into adding them here.
>
> I was thinking contract because it should be a mixing pot of different
> algorithms. Right now it only uses einsum; however, it will use vendor
> BLAS before it is complete and anything else that comes along can be added
> as well. For example, if numpy ever offered an interface to a tensor
> framework this could be leveraged as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5488#issuecomment-72656522>.
>

dgasmith commented:
@njsmith I was starting to thing along the same lines. The main issues that I see:
- A few extra keywords would need to be added `einsum` making this function appear more complex. I would guess at the minimum we would need:
   - memory - If 0 do not optimized the expression and either use BLAS or einsum, else allow optimization.
  - path - Can either be a string to choose the type of path optimization or a tuple which would represent a path to execute.
  - return_path - Returns a path that can be fed back into `einsum` and a human readable component.
- We would be increasing the overhead of the `einsum` function. Although, most of this should be moved to the C-side eventually.
- Optimization should probably be defaulted to off to prevent possible breaking of legacy code and unexpected memory overhead.

shoyer commented:
> Shouldn't the name be "einsum"?

Yes, of course.

If performance is a concern, my suggestion would be to write this in Cython, not C. That, along with caching, should take care of any performance concerns.

It doesn't particularly matter to me what form the path is returned in as long as it can be used as an argument to the function to skip the optimization step. Generating Python code is probably not a path we want to go down.

charris commented:
>Shouldn't the name be "einsum"? Names are for users, and users care about interfaces, not internal >implementation.

+1 as long as it doesn't break compatibility. With the long term prospect of moving more stuff down into multiarray, we might want to look into reorganizing that directory to separate out large, semi-standalone parts like this.

dgasmith commented:
I have been fairly busy this week, but I wanted to drop in and say that I am looking into this. I really like pushing everything into the `einsum` function and the extra control would be a nice addition.

---

@charris Can you expand a bit on reorganizing the `core` directory? I will likely be refactoring some of the `einsum` code anyway, so keeping this in mind would be helpful.

---

@shoyer Writing this in cython would be much easier, is the C conversion done with the [cythonize](https://github.com/numpy/numpy/blob/4cbced22a8306d7214a4e18d12e651c034143922/tools/cythonize.py) script? 

With the caching, we would need to do a thread safe implementation incase someone runs a threaded `einsum` implementation (e.g. `subprocess.Pool`). Do you have experience with this?

We may have missed each other at some point on the path saving functionality. Path saving is currently implemented and shown in the documentation. Is there anything wrong with the way it is done currently?

---

I probably need to run this through the mailing list again to solicit feedback. Overall, I think most of the major implementation questions have been answered.

shoyer commented:
@dgasmith I'm not entirely sure how the Cython build process is done in NumPy, but I know it *is* done (the `numpy.random` module is written in Cython).

This is actually a case where Python GIL makes things easy -- by default, everything in Python is thread safe. You simply would not release the GIL for the path optimization step.

Yes, I do see that path saving is current implemented. This is probably fine, though possibly even unnecessary if we get transparent caching going.

njsmith commented:
Unfortunately using cython here won't be as trivial as one might think. The
issue is that cython has a strong belief that each .pyx file should create
a separate shared library module. This is fine for np.random which lives
off on its own anyway and doesn't need any special access to numpy
internals, but the einsum implementation is currently part of the big
multiarray module.

Two possibilities for enabling cython usage:
1) pull all the einsum code out into its own module, like np.random. The
possible problem here is that if einsum relies on any internal functions
then it'll lose access to them. I don't know whether it does or not.

2) figure out how to coax cython into generating C code that can be built
into a single shared library with other code. This would be super useful
(there's lots of stuff in numpy that would be nicer to write in cython),
but requires some arcane knowledge of how cython and c and python all
interact.
On 9 Feb 2015 09:45, "Stephan Hoyer" <notifications@github.com> wrote:

> @dgasmith <https://github.com/dgasmith> I'm not entirely sure how the
> Cython build process is done in NumPy, but I know it *is* done (the
> numpy.random module is written in Cython).
>
> This is actually a case where Python GIL makes things easy -- by default,
> everything in Python is thread safe. You simply would not release the GIL
> for the path optimization step.
>
> Yes, I do see that path saving is current implemented. This is probably
> fine, though possibly even unnecessary if we get transparent caching going.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5488#issuecomment-73554094>.
>

dgasmith commented:
It seems just writing this in C would be the easiest route. The largest issue that I see is loosing out on set arithmetic. I would guess from the limited number of valid einsum indices (48) a bit array set implementation would be the easiest and likely very fast. Although, I have not thought about this before. Does anyone have an opinion before I dig into it?

juliantaylor commented:
a set of real world use cases where this is too slow would be useful first, to better judge what exactly needs to be speed up. Then we can decide how to do it.
I'd focus more on the features and tests, optimization can be done later.

for cython we could simple concatenate a bunch of files before running cython to get around the duplication issue.

juliantaylor commented:
from a brief look at the code it does not look like code that would profit much from using cython, it definitely needs to be profiled before deciding on the way to go. Unless there is one really obvious bottleneck that needs typing it looks like it should stay python or be really low level C using special purpose data structures instead of generic python ones.

shoyer commented:
Is there a straightforward way to use internal routines written in C for einsum (e.g., for argument parsing) on the Python side? That was part of the reason why I was thinking it might be desirable to do this from Cython/C.

jaimefrio commented:
In the `numpy.lib` module you have `np.interp`, which is a wrapper around a Python function written in C, `numpy.core.multiarray.interp` that does the heavy lifting. This C functions lives in `numpy/core/src/multiarray/compiled_base.c` and is made part of `numpy.core.multiarray` in `numpy/core/src/multiarray/multiarraymodule.c`. Unless we agree on a better place, I think `compiled_base.c` is the place to put non-core functionality to be used elsewhere.

As for writing stuff in Python first, right now `np.einsum` is defined in `numpy/core/numeric.py` to be `numpy.core.multiarray.einsum`, but we might as well add an `einsum` function there that did some preliminary work and ended up calling the C function afterwards.

dgasmith commented:
If next up is dropping this into `numpy/core/numeric.py` the first thing to do is likely the inevitable input parsing rewrite:
- Move the `numpy/core/src/multiarray/multiarraymodule.c` function `array_einsum` to `einsum_parse_input` and instead of calling `PyArray_EinsteinSum` (the actual einsum function) return a tuple or dictionary of the parsed input.
- Make a new wrapper in `multiarraymodule.c` to `PyArray_EinsteinSum` that takes the above dictionary or tuple.

I would prefer a dictionary over a tuple for passing information, but either is fine. My main concern would be ensuring that reference counting is still handled correctly.

If this sounds fine, I can try to write this out.

njsmith commented:
That sounds fine. Or instead of a dict/tuple (which are relatively
heavyweight python objects) it might make sense to just directly return
whatever c representation is most convenient to use downstream. (No point
in creating a fancy dict representation if we just end up having to "parse"
the dict later.)

But I haven't actually looked at the code, so for all I know it's already
using dicts internally :-)
On 11 Feb 2015 10:49, "Daniel Smith" <notifications@github.com> wrote:

> If next up is dropping this into numpy/core/numeric.py the first thing to
> do is likely the inevitable input parsing rewrite:
>
>    - Move the numpy/core/src/multiarray/multiarraymodule.c function
>    array_einsum to einsum_parse_input and instead of calling
>    PyArray_EinsteinSum (the actual einsum function) return a tuple or
>    dictionary of the parsed input.
>    - Make a new wrapper in multiarraymodule.c to PyArray_EinsteinSum that
>    takes the above dictionary or tuple.
>
> I would prefer a dictionary over a tuple for passing information, but
> either is fine. My main concern would be ensuring that reference counting
> is still handled correctly.
>
> If this sounds fine, I can try to write this out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5488#issuecomment-73939247>.
>

charris commented:
On a quick looksie, this looks sort of, but not quite, ready. In particular, the question of whether it should be an improved einsum, which requires complete compatibility. I'm tempted to put this off to Numpy 1.11.0 unless it becomes active again.

dgasmith commented:
@charris The code that actually does the optimization is fairly close to done. Where I keep running into issues is trying to take advantage of the current C parsing. We only really need the einsum string returned to the python side, unfortunately part of the parsing is written in the primary einsum code and is not easily extractable. I am not quite sure I can get the information back out without making fairly substantial changes to the code, which is not something I would really like to do.

The other option is for optimized einsum contractions to have its own parsing python side. If you need to worry about optimization you definitely will not notice the extra overhead. Although, it would create two codes that do the same thing.

charris commented:
Yeah, the einsum c parsing is a handful. I don't mind doing parsing python side. I guess the question is do we add a new function, or maybe a keyword. I wouldn't mind a bit of help here from the others who have looked at the code: @shoyer @njsmith @juliantaylor @jaimefrio .

njsmith commented:
From the user point of view I really think optimization should be as
transparent as possible. There's no reason to be shy about changing the
existing code -- the goal is to have one good einsum, not two that are good
at different pieces... maybe that means moving the parsing entirely into
python, e.g.

@charris <https://github.com/charris> The code that actually does the
optimization is fairly close to done. Where I keep running into issues is
trying to take advantage of the current C parsing. We only really need the
einsum string returned to the python side, unfortunately part of the
parsing is written in the primary einsum code and is not easily
extractable. I am not quite sure I can get the information back out without
making fairly substantial changes to the code, which is not something I
would really like to do.

The other option is for optimized einsum contractions to have its own
parsing python side. If you need to worry about optimization you definitely
will not notice the extra overhead. Although, it would create two codes
that do the same thing.

—
Reply to this email directly or view it on GitHub
<https://github.com/numpy/numpy/pull/5488#issuecomment-101435506>.

dgasmith commented:
@njsmith I was originally planning to leave optimization off by default. It is possible that turning optimization on would break some `einsum` use cases due to memory overhead. I believe this is an unlikely scenario, but worth considering.

dgasmith commented:
I started working on merging this into the einsum function with python side parsing:
https://github.com/dgasmith/numpy/tree/opt_einsum3

Thinking on this a bit more, all einsum kwargs besides `out` are invalid for optimized einsum and realistically optimization should only be attempted for arrays of `dtype=np.double`. It seems most of the changes to einsum make it more flexible, optimization will make it significantly less so.

njsmith commented:
It's fine if einsum is faster in simple cases and slower if more complex
features are used.
On May 19, 2015 6:18 PM, "Daniel Smith" <notifications@github.com> wrote:

> I started working on merging this into the einsum function with python
> side parsing:
> https://github.com/dgasmith/numpy/tree/opt_einsum3
>
> Thinking on this a bit more, all einsum kwargs besides out are invalid
> for optimized einsum and realistically optimization should only be
> attempted for arrays of dtype=np.double. It seems most of the changes to
> einsum make it more flexible, optimization will make it significantly less
> so.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5488#issuecomment-103711954>.
>

charris commented:
@dgasmith Keep us informed of progress. I might put this to numpy 1.11.

charris commented:
I've changed the milestone to the 1.11 release.

dgasmith commented:
Greetings all, I rewrote the einsum parsing python side, merged the optimization functions into einsum, and updated the test cases. The python side parsing is quite ugly, but I believe it does what its supposed to do. Please try to break it if you have the chance.

Looking through the code the only things left to do are update the documentation and figure out what to do with BLAS functionality. For the BLAS functionality it looks like the built-in BLAS is significantly slower than `einsum`, is there an easy way to detect if a library is linked besides `np.__config__.get_info('blas')`?

The history is pretty thoroughly messed up at this point. As all of the changes are fairly local its probably best to fix it with a new PR. For now the changes can be found in the following two files:

`numpy/core/tests/test_einsum.py`
`numpy/core/einsumfunc.py`

homu commented:
:umbrella: The latest upstream changes (presumably #6048) made this pull request unmergeable. Please resolve the merge conflicts.

dgasmith commented:
@homu I missed this when the error occurred and have now fixed the sync errors.

This has been dormant for a long time. Browsing through the code it appears close to merge ready; however, I am not sure what else needs to be changed without external feedback.

argriffing commented:
btw homu appears to be some kind of bot
https://github.com/barosl/homu

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+        current = new
+
+    # If we have not found anything return single einsum contraction
+    if len(new) == 0:
```
If input_sets has size 0 or 1, new could be undefined.

dgasmith replied:
I can add another conditional here. The `einsum` function is constructed in a way so that this cannot happen. As a general question should internal functions be fully protected in this way?

shoyer replied:
It's not necessary to protect internal functions like this, but if you skip checks like this, you should add a comment to note why it's safe.

charris replied:
+1 on @shoyer's comment.

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+        return [tuple(range(len(input_sets)))]
+
+    new.sort()
+    path = new[0][1]
```
Probably better just to use `min` instead of sorting.

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+
+    Returns
+    -------
+    if return_path:
```
Maybe it would be better to create another dedicated function `einsum_path` for computing the optimal path? It's an awkward API to make a function that returns something totally different if it gets a keyword argument.

dgasmith replied:
I like this idea a lot, I will move all of the path functionality into another function.

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+        - if a list is given uses this as the path.
+        - 'greedy' means a N^3 algorithm that greedyally
+            chooses the best algorithm.
+        - 'optimal' means a N! algorithm that tries all possible ways of
```
Maybe there is some cutoff for N at which it makes sense to switch behavior from optimal to greedy?

shoyer commented:
I'm sure someone has asked about this before, but remind me -- is there a reason you aren't using the [classic dynamic programming solution](https://en.m.wikipedia.org/wiki/Matrix_chain_multiplication#A_Dynamic_Programming_Algorithm) to this problem? That algorithm scales like O(n^3) for the exact solution.

shoyer commented:
@dgasmith Do you know any concrete examples where the optimal path construction is too slow compared to the greedy algorithm? That would be helpful for figuring out whether we need to optimize this further (e.g., by using dynamic programming).

One thing that I would like to see for this PR are explicit tests for the subroutines. It would be nice to test the optimization routines independently of einsum, for example, to be sure that they return reasonable results.

For the string summary of the optimization, it would be nice to include another column that shows the incremental number of operations for each reduction, along with the total cost. I might even use this instead of the "scaling" column -- it conveys the performance much more concrete.

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+einsum_symbols = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+einsum_symbols_set = set(einsum_symbols)
+
+def _compute_size_by_dict(indices, idx_dict):
```
Explicit doctest style examples for these helper functions would make it more obvious how they work

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+    path : bool or list, optional (default: ``greedy``)
+        Choose the type of path.
+
+        - if a list is given uses this as the path.
```
it would be good to explain how path works, possibly with an example, e.g., "a list of specific terms to contract in order"

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+    return ret
+
+
+def _find_contraction(positions, input_sets, output_set):
```
An example would be very helpful for this function

dgasmith replied:
Can do; however, this is simply a helper function. From reading the numpy code these types of functions typically do not have any documentation associated with them. Is there a good example I can view?

shoyer replied:
This is just my personal preference style. I like what @mrocklin does in dask, e.g., https://github.com/blaze/dask/blob/c4df983f03eeac02b5b99ba0628b2ead84b42a17/dask/array/core.py#L1204

In my opinion, examples can actually be better than full numpydoc style documentation of arguments. I was scratching my head a little bit to understand the difference between the "resulting" and "current" contraction :).

shoyer commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,597 @@
+                # Find cost
+                new_cost = _compute_size_by_dict(idx_contract, idx_dict)
+                if len(idx_removed) > 0:
+                    new_cost *= 2
```
Where does this times 2 come from?

dgasmith replied:
If there is a zip index as denoted by `idx_removed` the FLOP count increases by a factor of 2. This is not present in the greedy algorithm as contracts with zip indices are prioritized over all other operations.

shoyer replied:
Yep, I figured this out. I think a comment would be appropriate to clarify.

dgasmith commented:
A lot of your questions can be answered by referring to the longer documentation on the `opt_einsum` github page [here](https://github.com/dgasmith/opt_einsum). I was not sure how much of this to add to numpy, but your suggestions will serve as an excellent starting point.

Chain matrix multiplication is a special subset of this problem that can be solved exactly in `N^3` operations; however, general contraction optimization is factorial scaling. The greedy algorithm in many ways is a generalization of the dynamic chain matrix matrix multiplication algorithm that you linked.

For randomized inputs the greedy algorithm finds a path that is comparable to that of the optimal path in nearly all (read 99%+) of cases. The optimal path starts to become rather costly (~5ms at 5 terms). The cost of the optimal path can be reduced; however, for tractable gain this should really be pushed C side if the performance is really a concern. 

One thing to consider is that the acceptable overhead cost is greatly dependent on the cost of the contraction. So a definite line between optimal and greedy algorithms is hard to draw. As the performance of the greedy algorithm is so good it seemed simpler just to always pick this algorithm.

shoyer commented:
Yeah, I can see a case for only using the greedy approach. However, if there are very few cases where the performance differs, it might be better to use the exact strategy. In general I am slightly reluctant to expose an API for selecting the optimization strategy. This will not be relevant to 99% of users and precludes future improvements.

On Sat, Nov 28, 2015 at 8:08 AM, Daniel Smith <notifications@github.com>
wrote:

> A lot of your questions can be answered by referring to the longer documentation on the `opt_einsum` github page [here](https://github.com/dgasmith/opt_einsum). I was not sure how much of this to add to numpy, but your suggestions will serve as an excellent starting point.
> Chain matrix multiplication is a special subset of this problem that can be solved exactly in `N^3` operations; however, general contraction optimization is factorial scaling. The greedy algorithm in many ways is a generalization of the dynamic chain matrix matrix multiplication algorithm that you linked.
> For randomized inputs the greedy algorithm finds a path that is comparable to that of the optimal path in nearly all (read 99%+) of cases. The optimal path starts to become rather costly (~5ms at 5 terms). The cost of the optimal path can be reduced; however, for tractable gain this should really be pushed C side if the performance is really a concern. 
> One thing to consider is that the acceptable overhead cost is greatly dependent on the cost of the contraction. So a definite line between optimal and greedy algorithms is hard to draw. As the performance of the greedy algorithm is so good it seemed simpler just to always pick this algorithm.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/5488#issuecomment-160314855

shoyer commented:
I have some ideas for how to do a dynamic programming solution for the general contraction problem to bring the runtime down, but before I work on that I want to be sure it's actually relevant for some problems. Do you know specific cases where the runtime of the exact path optimization algorithm with memory bounds is a significant fraction of the runtime for the contraction itself?

dgasmith commented:
At 7 terms the exact path currently takes 270ms while the greedy path takes 190us; for 4 terms we see 250/45us. At the same time I have einsum function calls that take anywhere from 1us to several hours. The factorial scaling of the exact path is inherently limiting. 

Optimization beyond rank reduction is not really a concern right now as it is both difficult to do and only provides linear speed increases. Historically, the only reason the exact path was programmed was to prove that the greedy algorithm was correctly reducing the problem to the minimum possible rank.

shoyer commented:
@dgasmith I'm sure I'm missing something obvious, but it's still not clear to me why the dynamic programming approach fails. Is it because you need to look at all the subproblems to know when an index is being summed out? This would be useful to document :).

> At 7 terms the exact path currently takes 270ms while the greedy path takes 190us; for 4 terms we see 250/45us.

Is there a specific example that shows this? In all the toy benchmarks I set up, even with 7 terms, the performance was quite similar with the exact and greedy examples.

dgasmith commented:
@shoyer When you say dynamic programming approach, do you mean the formalism of caching intermediates or the chain matrix multiplication optimization neighboring pair routine itself?

For the former, you know quite a bit about what the input/ouput of chain multiplication will be beforehand making it quite easy to cast matrix multiplication in this formalism. We can certainly make the greedy and optimal expressions faster to evaluate, although I believe this would be difficult to do in python.

For the latter, take the following einsum expression `ac,abde,bc->de`. If we approach this like our typical matrix multiply problem by only considering neighboring pairs we get `(ac,abde),bc->de` which will scale like N^5 (the naive scaling). A better approach would be `(ac,bc),abcd->de` which scales like N^4. 

Looks like there was a bug that always caused the greedy path to evaluate. I have now fixed this.

To be clear I was comparing the cost of computing the *path* in the above example. The cost of the contraction will itself will likely be similar as the exact and greedy algorithms usually return the same result.

I have made some suggestions based on your comments. Still plenty of issues, but hopefully we are heading in the right direction.

homu commented:
:umbrella: The latest upstream changes (presumably #6630) made this pull request unmergeable. Please resolve the merge conflicts.

mhvk commented:
Saw mention of this on the mailing list. Would be *very* nice to have. It may not be crazy to default to optimizing; a memory overhead approximately equal to the largest item seems quite reasonable.

dgasmith commented:
@mhvk I would like to get this in as well. I was planning to go to the NumPy sprint at SciPy to finish it out, but it doesn't look like there will be one this year. 

I feel like this PR is difficult as its fairly large by NumPy standards and does some operations in slightly hacky ways (the duplicate python and C++ parsing for example). If we could have a discussion that figured out exactly what needed to be done and a consensus on default options I would be happy to do it and finished out this PR.

I would need to review the code, but from my standpoint I think the documentation is the primary issue left. Ill look at the code this weekend to see if there are any other issues that have crept up since the latest round of commits.

As a note some kind souls have converted the original [repository](https://github.com/dgasmith/opt_einsum) into a python package if you would like to use this separately.

charris commented:
@dgasmith I'm looking to get to work on this for 1.12. What is the current status? Have you had experience with the package? It would also be good if you could clean up the commit history a bit and adhere to the commit message format in `doc/source/dev/gitwash/development_workflow.rst`

dgasmith commented:
@charris The code is pretty well developed at this point with several production level codes depending on it. I think the largest outstanding issue is to see if the "tacking-on" of the optimization part of einsum in the manner shown [here](https://github.com/numpy/numpy/pull/5488/files#diff-8d3c9add8f7307f2b90bf7fca47f9908R683) is ok. The downside is the einsum input parsing is duplicated both in Python and C.

If this is fine I can work on another pass.

charris commented on numpy/core/tests/test_einsum.py:
```diff
@@ -5,281 +5,296 @@
-        # dtype parameter must be a valid dtype
-        assert_raises(TypeError, np.einsum, "", 0, dtype='bad_data_type')
+            # dtype parameter must be a valid dtype
+            assert_raises(TypeError, np.einsum, "", 0, dtype='bad_data_type', optimize=do_opt)
```
The long lines should be broken, note also that alignment should be at the opening parenthesis. Alternative would be to use `self.assertRaises` as a context manager, available in Python 2.7 and up.

charris commented on numpy/core/tests/test_einsum.py:
```diff
@@ -622,6 +668,104 @@ def test_small_boolean_arrays(self):
+        self.optimize_compare('ab,bcd,cd->abcd')
+        self.optimize_compare('ab,bcd,cd->abd') 
+
+    def test_previously_failed(self):
```
So these now work with all levels of optimization?

dgasmith replied:
Yes, these are previous test cases that were difficult for one reason or another. Perhaps this should be renamed.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,718 @@
+        Choose the type of path.
+
+        - if a list is given uses this as the path.
+        - 'greedy' A N^3 algorithm that chooses the best pair contraction
```
What is `N`?

dgasmith replied:
Number of terms in the contraction, will add a note.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,718 @@
+        Specifies the subscripts for summation.
+    *operands : list of array_like
+        These are the arrays for the operation.
+    tensordot : bool, optional (default: True)
```
Is there an advantage to `tensordot`?

dgasmith replied:
Not currently no; however, tensordot features can be added so that contractions like `abcd,cdef->abef` can utilize a BLAS call (GEMM in this case). The downside is tensordot always copies data and your np.dot protocol does extra copies as well for C-ordered arrays. There was never a decision if the speed vs memory overhead was worth it.

See [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/blas.py) for the BLAS lines that would be added.

charris replied:
If not used, it should probably be removed from the Parameters. It can always be added later.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,718 @@
+    Notes
+    -----
+    Subscript labels follow the same convention as einsum with the exception
+    that integer indexing and ellipses are not currently supported.
```
Is it still the case that ellipses are not supported? I see that they are parsed.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,718 @@
+    return (path, path_print)
+
+# Rewrite einsum to handle different cases
+def einsum(*operands, **kwargs):
```
I rather prefer the `c_einsum` signature
```
einsum(subscripts, *operands, out=None, dtype=None, order='K', casting='safe')
```

dgasmith replied:
I don't think this signature is possible in python. We could do:
```python
einsum(subscripts, out=None, dtype=None, order='K', casting='safe', *operands)
```

shoyer replied:
It works in Python 3, which has keyword only arguments.

To get something working in Python 2, you have to unpack `kwargs` manually, e.g.,
```
def einsum(subscripts, *operands, **kwargs):
    out = kwargs.pop('out', None)
    dtype = kwargs.pop('dtype', None)
    ...
    if kwargs:
        raise TypeError('einsum() got unexpected keyword arguments: %s'
                        % list(kwargs))
```

charris replied:
Hmm, looks like that `einsum` signature was due to the documentation rather than the C function and is something implemented at the c level, so there isn't much choice here. I guess the question becomes whether to check for unexpected keyword arguments.

dgasmith committed:
ENH: Allows contraction order optimization in einsum function.

dgasmith commented:
@charris, @shoyer I have pushed a refactored version of the code based off the current opt_einsum repository. The code base is largely the same and incorporates all previous feedback. New features:

- All tensordot functionality has been removed for the time being.
- The documentation has been cut temporarily as the optimization feature should just be a note on the current documentation.
- New `einsum_path` function to view and print a human readable version of the path along with theoretical speed up, scaling reduction, and largest intermediate size. Most of the important optimization documentation now lives here.
- New python-side parsing now supports all available c-side parsing functionality.
- Explicit path testing.
- Explicit kwarg testing.

Ill fix the long line and documentation issues on the next pass.

dgasmith commented on numpy/core/tests/test_einsum.py:
```diff
@@ -1,284 +1,301 @@
-        assert_raises(TypeError, np.einsum, "", 0, bad_arg=0)
+            # other keyword arguments are rejected
+            # Running this is python will raise this error
+            #assert_raises(TypeError, np.einsum, "", 0, bad_arg=0, optimize=do_opt)
```
If I simply call the above from `numpy` I get the `TypeError` as expected. However, when run with nose it doesn't see a type error raised and the test case fails. Any ideas?

seberg replied:
I am only getting the type error when optimize is True, not when it is False. The test seems to check both.

dgasmith replied:
@seberg Yup, thanks. Was cleaning the variables too early.

seberg commented on numpy/core/tests/test_einsum.py:
```diff
@@ -1,284 +1,301 @@
 from __future__ import division, absolute_import, print_function
 
+import warnings
```
Adding the import should be unnecessary?

dgasmith replied:
Apparently not. Not entirely sure where that came from.

charris commented:
Somewhere along the line you should add something to `doc/release/1.12.0-notes.rst`. I think this counts as one of the `Highlights`. The notes are currently in very rough form so don't worry too much about that.

dgasmith commented:
@charris Gave the documentation a pass and pep8ified the code.

Two remaining questions:
- Should `einsum_path` just be an argument to `einsum`? Right now I think it helps as we can move a considerable amount of path information to this function.
- Should `optimize` be turned on or off by default? Ive tested a few million possible inputs with 3-6 operands and everything works fine. So I think the path code is reasonably bulletproof. On the other hand for people who work with very large arrays it could potentially change any memory limit checks that they have in place.

charris commented:
@dgasmith I think the main consideration in making optimization the default is execution time for small matrices and simple contractions.

Could you explain the first question a bit. I'm not clear on what you are proposing. I'm also not sure removing the add_doc documentation is the right thing. The `multiarray.einsum` function still exists and should be documented. Could add a note to the documentation pointing out the difference.

I'm not convinced `einsumfunc.py` is a better name than the simple `einsum.py`, but don't feel it is worth worrying much about. It is easy enough the change down the line if desired.

dgasmith commented:
@charris The complete overhead for computing a path (parsing the input, finding the path, and organization that data) with default options is about 150us. Looks like einsum takes a minimum of 5-10us to call as a reference. So the worst case scenario would be that the optimization overhead makes einsum 30x slower. Personally id go for turning optimization off by default and then revisiting if someone tackles the parsing issue to reduce the overhead.

For my first question I added a function called `np.einsum_path` that prints out a detailed map of what an optimized path does. This can be rolled into `np.einsum` by adding something like `return_path` to `np.einsum` itself. I don't know how sensitive you guys are about adding to the base namespace.

Should I revert the add_doc documentation, but leave the documentation in `einsumfunc.py`? Not quite sure what your are detailing here.

I don't really have an opinion on most documentation/filename issues. Just tell me what needs fixing.

charris commented:
Given the call overhead, I agree that turning off optimization by default is the way to go. 

Hmm, the examples `add_doc` are correct but now refer to the "new" einsum. I note that it would be very easy to rename `multiarray.einsum`, just change the exported name in `multiarraymodule.c`. Maybe call it `"c_einsum"` or even `"_einsum"`. Note that I am just tossing out suggestions here as having identical names for two different functions, both of which we would like to document, just seems awkward.

I am not a fan of multiple outputs depending on passed parameters. It seems simple enough to call `np.einsum_path` for the curious who want to explore just how neat it is :)

@shoyer @njsmith @seberg This is getting close to a merge, I'd welcome any final input at this point.

charris commented:
Note that at some future date it might make sense to call `einsum_path` from the C code and put the rest of the argument parsing and contraction loop in `array_einsum`. I think moving things around like that in the future would not be a big compatibility problem and might cut down on the call overhead.

seberg commented:
If it is very simple, another choice might be to default to no optimization based on some simple heuristic such as the input size. In that case it could be even neater to call back into python for optimization instead the other way around. However, I guess it should not make a big difference.

The einsum path information seems useful if you have a bigger problem and want to know how to approach it. I could also imagine just makeing it `np.lib.einsum.einsum_path` (or whatever) and pointing to it from the documentation, though.

Have to look more at this to give real input though, will try to do that later.

juliantaylor commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,967 @@
+        Also accepts an explicit contraction list from the ``np.einsum_path``
+        function.
+        See ``np.einsum_path`` for more details.
+    memory_limit : int, optional
```
a parameter that only has an effect if another a parameter is set is not very nice
maybe accepting a tuple for optimize with the first being the optimize value and the second the memory limit is better?

dgasmith replied:
Sure, we can do that. Id like to make it so that someone can simply say `optimize=True` as well. The default `memory_limit` is pretty reasonable and is fine for 99% of cases I think.

juliantaylor commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,967 @@
+
+    See Also
+    --------
+    einsum_path, dot, inner, outer, tensordot
```
a reference to linalg.multi_dot would be nice as it is very related for the chained dot case

dgasmith commented:
@seberg The biggest problem I've had with simple heuristics is parsing the input python-side is quite slow by itself (30-100us), at that point computing the path isn't a huge overhead. Parsing C-side is really quite beneficial, disentangling einsum from its parsing C-side seemed like a fairly large project however.

One possibility would be to add a series of heuristics that check at different parsing stages if you should default back to c_einsum or not. This has always seemed more complex than its worth at this stage.

I liked the idea of removing memory_limit as a separate keyword so I rearranged that a bit.

charris commented:
@dgasmith Let's rename `"einsum"`, line 4142 in `multiarraymodule.c` to `"_einsum"` and fix the import statement in `numpy/core/einsumfunc.py`. The original documentation with a modification mentioning that the function private and a see also reference to `np.einsum` should be good enough.

charris commented:
@dgasmith I think we can put this in once the remaining details are taken care of.

homu commented:
:umbrella: The latest upstream changes (presumably #7980) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+from numpy.core.multiarray import einsum as c_einsum
```
This need some module documentation and the standard lead in
```
"""
Implementation of optimized einsum.

"""
from __future__ import division, absolute_import, print_function

```

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+    overall_contraction = input_subscripts + "->" + output_subscript
+    header = ("scaling", "current", "remaining")
+
+    speedup = naive_cost / float(opt_cost)
```
With Python 3 division (see above), the float is not needed.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+def einsum(*operands, **kwargs):
+    """
+    einsum(subscripts, *operands, out=None, dtype=None, order='K',
+           casting='safe', optimize=True, memory_limit=None)
```
The optimize default is `False`, not `True`.

charris replied:
`memory_limit` seems to be gone.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+        Specifies the subscripts for summation.
+    operands : list of array_like
+        These are the arrays for the operation.
+    out : ndarray, optional
```
`{ndarray, None}`

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+        These are the arrays for the operation.
+    out : ndarray, optional
+        If provided, the calculation is done into this array.
+    dtype : data-type, optional
```
This should be `{data-type, None}`.

charris replied:
Also needs the default value documented.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+          * 'safe' means only casts which can preserve values are allowed.
+          * 'same_kind' means only safe casts or casts within a kind,
+            like float64 to float32, are allowed.
+          * 'unsafe' means any data conversions may be done.
```
Need default documented.

dgasmith replied:
@charris I could not find another place were casting had a default or where there was extra text after a list. So I hope the following is parsed correctly. Should the default be added to other casting descriptions as well?

charris replied:
The default is `'safe'`, implemented in `array_einsum`. Not sure what to say for the `dtype` argument, maybe smallest type compatible with safe casting.

We try to document the default values of all optional arguments, although we evidently haven't been perfect about that. So the answer to the last guestion is "yes".

shoyer replied:
In my opinion it's actually a cleaner style to document default values only in function signatures. Otherwise, it's easy to get out of sync.

charris commented on numpy/core/einsumfunc.py:
```diff
@@ -0,0 +1,979 @@
+        Controls if intermediate optimization should occur. No optimization
+        will occur if False and True will default to the 'greedy' algorithm.
+        Also accepts an explicit contraction list from the ``np.einsum_path``
+        function.
```
Needs default documented. I wouldn't put `See ...` on a separate line.

charris replied:
What happened to `memory_limit`, it is still part of the function signature.

dgasmith replied:
I popped the `memory_limit` argument as @juliantaylor recommended. I updated the documentation, but didnt take out `memory_limit` in other places. Ill fix this up as well.

charris commented:
The conflict is the release notes. Three things that need fixing/checking
- Compatibility header at beginning of `einsumfunc`.
- Make sure documented signatures are correct.
- Optional parameters need to have the default value listed.

Changing my previous suggestion, `c_einsum` instead of `_einsum` for the multiarray version. The original documentation can be obtained from master with `git co master numpy/add_newdocs.py` and edited for the new name.

If you don't have time for this, let me know and I will fix up the few remaining nits.

dgasmith commented:
@charris Ill clean up the remaining issues Monday morning. Thanks for looking through this!

charris commented on numpy/add_newdocs.py:
```diff
@@ -2150,14 +2152,16 @@ def luf(lamdaexpr, *args, **kwargs):
+    einsum, dot, inner, outer, tensordot
 
     Notes
     -----
```
IIRC, the examples need changing, but I can do that later. I expect this function might be changing a bit down the line in any case.

charris commented:
OK, let's get this in before the release notes change again. Thanks @dgasmith.

dgasmith commented:
@charris Great, thanks for getting this merged in! Please ping me for any required fixes.

I will probably have some tweaks to the algorithms down the line and ill probably look into adding BLAS calls back as well.

shoyer commented:
This is awesome, thanks @dgasmith for all your hard work on this one!
